google/cloud/pubsub_v1/subscriber/_protocol/requests.py

```diff
@@ -39,9 +44,11 @@ class LeaseRequest(NamedTuple):
 class ModAckRequest(NamedTuple):
     ack_id: str
     seconds: float
+    future: Optional["futures.Future"]
```

Comment:
Where is this implemented? Sorry I am struggling to find it in the CL.

Reply:
Whether a future is set or not depends on whether the user calls ack or ack_with_response (or the modack & nack equivalents). This is done in `google/cloud/pubsub_v1/subscriber/message.py`.

If a future is present, we put it in an ack_id->future in dispatcher.py and pass it to the streaming_pull_manager (send_unary_ack or send_unary_modack methods). Those methods are responsible for finishing the futures or returning failed ones to be retired by the dispatcher.